### ISSUE TITLE
UTF8 support?

### ISSUE THREAD

**Issue by joakim-tjernlund** [troglobit/editline#37]
Any idea if UTF support can be added to this lib?
Preferably without depending on locale stuff from glibc

**Comment by troglobit** [comment]
Sure it can, it's just a lot of work since most of the lib is built around the notion that a character is one byte wide. A Unicode character, as far as I know, can be two, three, or four bytes wide.

Do you mind telling med the usecase? If our interests interlock (which I've noticed quite a few times) I may be able to help in such an endeavor.

**Comment by joakim-tjernlund** [comment]
Today most terminals default to UTF-8 so i think it wold be great if editline could do UTF-8
but limit accepted chars to ISO Latin  .
That would help our use case where users can connect via telnet/ssh/rs232 and non ascii
chars would just work.

**Comment by troglobit** [comment]
Yeah, that's a good limitation. I'll have a look again and try to estimate the amount of work needed.

**Comment by troglobit** [comment]
Interestingly quite a lot actually works already.  What needs attention is cursor movement, most notably when editing a line.  It is hard-coded to one-character=one-byte ... also, most of the internals use 'int' to store a character during processing.  So it shouldn't be too much work to straighten out libeditline \o/

**Comment by joakim-tjernlund** [comment]
Any progress on this?

**Comment by troglobit** [comment]
Nope. I've been redirected/reprioritized at work atm, so I'll have to circulate back to this later. Sorry.

**Comment by steakknife** [comment]
The easiest way to do this is like ncurses: make a slightly different API (functions, header and library) with a suffix like `w`. Then, use something like libunistring or icu4c to handle conversion from bytes to iterable characters th.

**Comment by troglobit** [comment]
@steakknife Thank you for the input!  I sort of came to the same conclusion wrt parsing, but I'm really hesitant to make editline depend on other libraries, the whole point of editline is to be small and have as few (no) dependencies as possible.  Whenever I circle back to this (or someone else does), your input will be a valuable starting point for further investigation!

**Comment by Artoria2e5** [comment]
I doubt there is any need for an extra layer on top of that, let along any libraries. The one thing we need to get from a UTF-8 string is the width of characters. Every POSIX system has a `int wcwidth(wchar_t c)`, and all we need is to parse UTF-8 into individual code points to feed to it. (On Windows you will need to carry your own `wcwidth` table. [busybox](https://github.com/mirror/busybox/blob/master/libbb/unicode.c) has a nice slim version.)

**Comment by troglobit** [comment]
Good point. Just remember that BusyBox is GPL, this lib isn't.

**Comment by minfrin** [comment]
Replxx (BSD license) has an example of calculating the number of unicode codepoints in a UTF8 string:

https://github.com/AmokHuginnsson/replxx/blob/master/examples/util.c#L3

**Comment by timkuijsten** [comment]
When it comes to supporting different locales I found this presentation by the OpenBSD maintainer of [libedit](https://cvsweb.openbsd.org/cgi-bin/cvsweb/src/lib/libedit/) very helpful: https://czyborra.com/yudit/eurobsdcon2016-utf8.pdf

Supporting UTF-8 might be possible without having to duplicate all interfaces or dragging in heavy weights like libiconv.

**Comment by steakknife** [comment]
@troglobit To be honest, Unicode (UTF-8, -16LE/BE, -32LE/BE, and UCS2/4) support including BOM detection/removal, sanitization, and normalization is trivial-adjacent. The catch is 95% of libraries get it wrong and/or don't keep it current, including tools at work used to parse Rust code. The maintenance required is precision around keeping current with Unicode versions and correctly classifying characters that could fall under multiple domains. The older Unicode references list these concerns more pedantically than newer ones.

**Comment by steakknife** [comment]
> @minfrin 

Nice. I recall having to fork [hirb-unicode](https://github.com/steakknife/hirb-unicode) to correctly draw text tables in an ActiveRecord ORM containing Hebrew and Japanese.

```
Input: 
   unsigned integer c - the code point of the character to be encoded
Output: 
   byte b1, b2, b3, b4 - the encoded sequence of bytes
Algorithm:
   if (c<0x80) 
      b1 = c>>0  & 0x7F | 0x00
      b2 = null
      b3 = null
      b4 = null
   else if (c<0x0800)
      b1 = c>>6  & 0x1F | 0xC0
      b2 = c>>0  & 0x3F | 0x80
      b3 = null
      b4 = null
   else if (c<0x010000)
      b1 = c>>12 & 0x0F | 0xE0
      b2 = c>>6  & 0x3F | 0x80
      b3 = c>>0  & 0x3F | 0x80
      b4 = null
   else if (c<0x110000)
      b1 = c>>18 & 0x07 | 0xF0
      b2 = c>>12 & 0x3F | 0x80
      b3 = c>>6  & 0x3F | 0x80
      b4 = c>>0  & 0x3F | 0x80
   end if
```

https://herongyang.com/Unicode/UTF-8-UTF-8-Encoding-Algorithm.html

Absolute range of U+0 to U+10FFFF

UTF-16 surrogate pairs are U+D800 to U+DFFF.

What is valid depends on the Unicode version. Validation code can be mechanically generated from a list of all valid code points. [Here's the latest database.](https://www.unicode.org/Public/UCD/latest/ucdxml/ucd.all.flat.zip)

**Comment by Artoria2e5** [comment]
Well uh... I don't think a line-editing program needs to concern itself with BOM, or to know about *any* Unicode properties, let alone hard-code ranges of assigned UTF code points. Heck, do we even need to check for surrogates for "valid" UTF-8 as opposed to WTF-8? If someone types it it's their fault, and even that could have use on quirky filesystems (looks at Windows). Grapheme editing is also overrated -- being able to navigate individual code points in sequences is actually useful!

**Comment by troglobit** [comment]
Whoa, a sudden burst of activity in this old thread! :smiley:

Lots of interesting suggestions coming in, so let me just reiterate: no external dependencies, and keep it simple.  No need to go full-blown unicode support, it's fine to start with a release that only does utf-8 and possibly even having to explicitly enable it at build time.
